File: BFSalgo/grid.cpp
#include "grid.hpp"

Grid2D::Grid2D(int w, int h, Cell defaultCell)
    : width(w), height(h),
        cells(h, std::vector<Cell>(w, defaultCell)) {}

void Grid2D::set(int x, int y, Cell value) {
    if (inBounds(x, y)) cells[y][x] = value;
}

Cell Grid2D::get(int x, int y) const {
    if (inBounds(x, y)) return cells[y][x];
    return Cell::UNKNOWN;
}

bool Grid2D::inBounds(int x, int y) const {
    return x >= 0 && x < width && y >= 0 && y < height;
}

void Grid2D::display() const {
    for (int y = 0; y < height; y++) {
        for (int x = 0; x < width; x++) {
            switch (cells[y][x]) {
                case Cell::FREE:        std::cout << ". "; break;
                case Cell::OBSTRUCTED:  std::cout << "# "; break;
                case Cell::UNKNOWN:     std::cout << "? "; break;
            }
        }
        std::cout << '\n';
    }
}

File: pyNav/grid.py
from enum import IntEnum

class Cell(IntEnum):
    UNKNOWN = 0
    FREE = 1
    OBSTRUCTED = 2

class Grid2D:
    def __init__(self, width: int, height: int, defaultCell: Cell = Cell.UNKNOWN):
        self.width = width
        self.height = height
        self.cells = [[defaultCell for _ in range(width)] for _ in range(height)]

    def set(self, x: int, y: int, value: Cell):
        if self.inBounds(x, y):
            self.cells[y][x] = value

    def get(self, x: int, y: int) -> Cell:
        if self.inBounds(x, y):
            return self.cells[y][x]
        return Cell.UNKNOWN

    def inBounds(self, x: int, y: int) -> bool:
        return 0 <= x < self.width and 0 <= y < self.height

    def display(self):
        for row in self.cells:
            for cell in row:
                if cell == Cell.FREE:
                    print('. ', end='')
                elif cell == Cell.OBSTRUCTED:
                    print('# ', end='')
                else:
                    print('? ', end='')
            print()